internal/proxyservice/proxy_service.go

```diff
@@ -57,89 +111,145 @@ func (s *proxyServices) Stop() error {
+
+	var subservices []httpproxy.Subservice
+	for _, subCfg := range httpSvcCfg.SubserviceConfigs {
+		// "cur" naming prefix needed to avoid package name collision
+		curPlugin := plugins[subCfg.Connector]
+		connResources := s.connectorResources(subCfg)
+		curConnector := curPlugin.NewConnector(connResources)
+		credsRetriever := s.credsRetriever(subCfg.Credentials)
+
+		subservices = append(subservices, httpproxy.Subservice{
+			Connector:           curConnector,
+			RetrieveCredentials: credsRetriever,
+		})
 	}
 
-	//TODO: Add validation to only support expected socket types
-	//TODO: Add validation somewhere about overlapping listenOns
-	listener, err := net.Listen(addressParts[0], addressParts[1])
+	// Create the logger
+	// HTTP proxy service gets its own logger (subservices have own loggers)
+
+	proxyName := httpSvcCfg.Name()
+	svcLogger := s.loggerFor(proxyName)
+
+	newSvc, err := httpproxy.NewProxyService(subservices, listener, svcLogger)
 	if err != nil {
-		s.logger.Errorf("could not create listener on: %s", svc.ListenOn)
+		s.logger.Errorf("could not create http proxy service '%s'", proxyName)
 		return nil, err
 	}
+	return newSvc, nil
+}
 
-	svcLogger := s.logger.CopyWith(svc.Name, s.logger.DebugEnabled())
-	connResources := connector.NewResources(svc.ConnectorConfig, svcLogger)
-	svcConnector := plugin.NewConnector(connResources)
+func (s *proxyServices) createTCPService(
+	config v2.Service,
+	pluginInst tcp.Plugin,
+) (internal.Service, error) {
 
-	// Temp var required so that the function closes over the current
-	// loop value.
-	credsCopy := svc.Credentials
-	credsRetriever := func() (map[string][]byte, error) {
-		return GetSecrets(credsCopy)
+	// TODO: Add validation somewhere about overlapping listenOns
+	// TODO: v2.NetworkAddress is a value type.  It needs to be moved to its
+	//   own package with no deps (stdlib deps are ok).
+	netAddr := v2.NetworkAddress(config.ListenOn)
+	listener, err := net.Listen(netAddr.Network(), netAddr.Address())
+	if err != nil {
+		s.logger.Errorf("could not create listener on: %s", config.ListenOn)
+		return nil, err
 	}
 
+	connResources := s.connectorResources(config)
+	svcConnector := pluginInst.NewConnector(connResources)
+	credsRetriever := s.credsRetriever(config.Credentials)
+
 	newSvc, err := tcpproxy.NewProxyService(
 		svcConnector,
 		listener,
-		svcLogger,
+		connResources.Logger(),
 		credsRetriever,
 	)
 
 	if err != nil {
-		s.logger.Errorf("could not create proxy service '%s'", svc.Name)
+		s.logger.Errorf("could not create proxy service '%s'", config.Name)
 		return nil, err
 	}
 
 	return newSvc, nil
 }
 
+func (s *proxyServices) connectorResources(svc v2.Service) connector.Resources {
+	svcLogger := s.loggerFor(svc.Name)
+	return connector.NewResources(svc.ConnectorConfig, svcLogger)
+}
+
+func (s *proxyServices) loggerFor(name string) logapi.Logger {
+	return s.logger.CopyWith(name, s.logger.DebugEnabled())
+}
+
+func (s *proxyServices) credsRetriever(
```

Comment:
proxyServices seems super busy. there's a lot of responsibility here. nothing comes to mind immediately for how to break it up but just wanted to bring that up. i'll be thinking about it ...

Reply:
let's discuss this afternoon

Reply:
closed per discussion on phone